File: 5-Problem-A-Day/107. Binary Tree Level Order Traversal II.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        LinkedList<List<Integer>> ans = new LinkedList<>();
        if (root == null)
            return ans;

        Queue<TreeNode> nodes = new LinkedList<>();
        Queue<Integer> levels = new LinkedList<>();
        nodes.add(root);
        levels.add(1);
        while (nodes.size() > 0) {
            TreeNode node = nodes.remove();
            int level = levels.remove();
            if (level > ans.size())
                ans.add(new LinkedList<>());
            ans.getLast().add(node.val);
            if (node.left != null) {
                nodes.add(node.left);
                levels.add(level + 1);
            }
            if (node.right != null) {
                nodes.add(node.right);
                levels.add(level + 1);
            }
        }
        Collections.reverse(ans);
        return ans;
    }
}

File: 2021/107. Binary Tree Level Order Traversal II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        if (root == NULL) return {};
        stack<vector<int>> nums;
        queue<TreeNode*> queue;
        queue.push(root);
        while (!queue.empty()) {
            int size = queue.size();
            vector<int> oneLevel;
            for (int i = 0; i < size; i++) {
                root = queue.front();
                queue.pop();
                oneLevel.push_back(root->val);
                if (root->left != NULL) queue.push(root->left);
                if (root->right != NULL) queue.push(root->right);
            }            
            nums.push(oneLevel);
        }
        vector<vector<int>> ans;
        while (!nums.empty()) {
            ans.push_back(nums.top());
            nums.pop();
        }
        return ans;
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        if (root == NULL) return {};
        vector<vector<int>> res;
        stack<pair<TreeNode*, int>> stack;
        stack.push({root, 1});        
        while (!stack.empty()) {
            root = stack.top().first;
            int level = stack.top().second;
            stack.pop();
            if (level > res.size()) res.push_back({});
            res[level - 1].push_back(root->val);            
            if (root->right) stack.push({root->right, level + 1});
            if (root->left) stack.push({root->left, level + 1});
        }
        reverse(res.begin(), res.end());
        return res;
    }
};